packages/angular_devkit/build_webpack/src/app-shell/index.ts

```diff
@@ -0,0 +1,181 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  BuildEvent,
+  Builder,
+  BuilderConfiguration,
+  BuilderContext,
+} from '@angular-devkit/architect';
+import { Path, join, normalize, virtualFs } from '@angular-devkit/core';
+import { Observable } from 'rxjs/Observable';
+import { forkJoin } from 'rxjs/observable/forkJoin';
+import { fromPromise } from 'rxjs/observable/fromPromise';
+import { merge } from 'rxjs/observable/merge';
+import { of } from 'rxjs/observable/of';
+import { _throw } from 'rxjs/observable/throw';
+import { concatMap, map, switchMap } from 'rxjs/operators';
+import { requireProjectModule } from '../angular-cli-files/utilities/require-project-module';
+import { BuildWebpackServerSchema } from '../server/schema';
+import { BuildWebpackAppShellSchema } from './schema';
+
+
+export class AppShellBuilder implements Builder<BuildWebpackAppShellSchema> {
+
+  constructor(public context: BuilderContext) { }
+
+  run(builderConfig: BuilderConfiguration<BuildWebpackAppShellSchema>): Observable<BuildEvent> {
+    const options = builderConfig.options;
+
+    return new Observable<BuildEvent>(obs => {
+      let success = true;
+      const subscription = merge(
+        this.build(options.serverTarget, {}),
+        this.build(options.browserTarget, { watch: false }),
+      ).subscribe((event: BuildEvent) => {
+        // TODO: once we support a better build event, add support for merging two event streams
+        // together.
+        success = success && event.success;
+      }, error => {
+        obs.error(error);
+      }, () => {
+        obs.next({ success });
+        obs.complete();
+      });
+
+      // Allow subscriptions to us to unsubscribe from each builds at the same time.
+      return () => subscription.unsubscribe();
+    }).pipe(
+      switchMap(event => {
+        if (!event.success) {
+          return of(event);
+        }
+
+        return this.renderUniversal(options);
+      }),
+    );
+  }
+
+  build(targetString: string, overrides: {}) {
+    const architect = this.context.architect;
+    const [project, target, configuration] = targetString.split(':');
+
+    // Override browser build watch setting.
+    const builderConfig = architect.getBuilderConfiguration<{}>({
+      project,
+      target,
+      configuration,
+      overrides,
+    });
+
+    return architect.getBuilderDescription(builderConfig).pipe(
+      concatMap(description => architect.validateBuilderOptions(builderConfig, description)),
+      switchMap(config => architect.run(config, this.context)),
```

Comment:
With a `BuilderConfiguration` you can just call `architect.run(config, this.context)`, you don't need to go validate the options manually. `.run` will do that for you.

Reply:
Do you mean with a description?

Reply:
No, I really do mean with a builder config. This is very similar to what the Architect CLI is doing: https://github.com/angular/devkit/blob/bee027fdd9b2d3d8ddc8c56ff487c9d2b641a6c4/packages/angular_devkit/architect_cli/bin/architect.ts#L130-L132